pylinalg/func/vector.py

```diff
@@ -116,3 +116,195 @@ def vector_apply_matrix(vectors, matrix, /, *, w=1, out=None, dtype=None):
         out = out.astype(dtype, copy=False)
     return out
+
+
+def vector_unproject(vector, projection_matrix, /, *, depth=0, out=None, dtype=None):
+    """
+    Un-project a vector from 2D space to 3D space.
+
+    Find a ``vectorB`` in 3D euclidean space such that the projection
+    ``projection_matrix @ vectorB`` yields the provided vector (in 2D euclidean
+    space). Since the solution to the above is a 1D subspace of 3D space (a
+    line), ``depth`` is used to select a single vector within.
+
+    Parameters
+    ----------
+    vector : ndarray, [2]
+        The vector to be un-projected.
+    projection_matrix: ndarray, [4, 4]
+        The camera's intrinsic matrix.
+    depth : number, optional
+        The distance of the unprojected vector from the camera.
+    out : ndarray, optional
+        A location into which the result is stored. If provided, it
+        must have a shape that the inputs broadcast to. If not provided or
+        None, a freshly-allocated array is returned. A tuple must have
+        length equal to the number of outputs.
+    dtype : data-type, optional
+        Overrides the data type of the result.
+
+    Returns
+    -------
+    projected_vector : ndarray
+        The unprojected vector in 3D space
+    """
+
+    raise NotImplementedError()
+
+
+def vector_apply_quaternion_rotation(vector, quaternion, /, *, out=None, dtype=None):
+    """Rotate a vector using a quaternion.
+
+    Parameters
+    ----------
+    vector : ndarray, [3]
+        The vector to be rotated.
+    quaternion : ndarray, [4]
+        The quaternion to apply (in xyzw format).
+    out : ndarray, optional
+        A location into which the result is stored. If provided, it
+        must have a shape that the inputs broadcast to. If not provided or
+        None, a freshly-allocated array is returned. A tuple must have
+        length equal to the number of outputs.
+    dtype : data-type, optional
+        Overrides the data type of the result.
+
+    Returns
+    -------
+    rotated_vector : np.ndarray
+        The rotated vector.
+
+    """
+
+    raise NotImplementedError()
+
+
+def vector_spherical_to_euclidean(spherical, /, *, out=None, dtype=None):
+    """Convert spherical -> euclidian coordinates.
+
+    Parameters
+    ----------
+    spherical : ndarray, [3]
+        A vector in spherical coordinates (r, phi, theta).
+    out : ndarray, optional
+        A location into which the result is stored. If provided, it
+        must have a shape that the inputs broadcast to. If not provided or
+        None, a freshly-allocated array is returned. A tuple must have
+        length equal to the number of outputs.
+    dtype : data-type, optional
+        Overrides the data type of the result.
+
+    Returns
+    -------
+    euclidean_coordinates : ArrayLike
+        A vector in euclidian coordinates.
+
+    """
+
+    raise NotImplementedError()
+
+
+def vector_distance_between(vectorA, vectorB, /, *, out=None, dtype=None):
```

Comment:
@Korijn This function is `np.linalg.norm(vectorA-vectorB)`. Is it worth keeping?

Reply:
Maybe it's still worth for it's semantics - code that uses it becomes easier to read?

Reply:
You mean when you use `np.subtract` with `out` to perform the subtraction in place? That could work, but only if we know that we can overwrite one of the two input vectors.

I ran a profile where I compared `np.subtract` with `out` against the naive approach, and in-place is better for small arrays, but for large arrays `np.linalg.norm` appears to dominate runtime:

![distance_timings](https://user-images.githubusercontent.com/4402489/207579086-357421b6-7a35-4031-afd3-8424a845f4ab.png)

<details>
<summary>Code</summary>

```python
import numpy as np
import matplotlib.pyplot as plt
from timeit import timeit

rng = np.random.default_rng(42)
matrix_size = np.linspace(10, int(1e7), 100).astype(int)

vectorA = [
    rng.random((x, 3))
    for x in matrix_size
]

vectorB = [
    rng.random((x, 3))
    for x in matrix_size
]

n_repeats = 5

naive_times = []
optimized_times = []

for idx in range(len(matrix_size)):
    naive_times.append(
        timeit(
            "np.linalg.norm(vectorA[idx]-vectorB[idx], axis=-1)",
            globals=globals(),
            number=n_repeats,
        )
        / n_repeats
    )
    optimized_times.append(
        timeit(
            "np.linalg.norm(np.subtract(vector, vectorB[idx], out=vector), axis=-1)",
            setup="vector = vectorA[idx].copy()",
            globals=globals(),
            number=n_repeats,
        )
        / n_repeats
    )

plt.plot(matrix_size, naive_times, label="Naive")
plt.plot(matrix_size, optimized_times, label="Inplace")
plt.gca().set_title("Performance (log-log)")
plt.gca().set_xlabel("Number of Vectors")
plt.gca().set_ylabel("Time per run")
plt.xscale("log")
plt.yscale("log")
plt.legend(title="Method")
plt.savefig("distance_timings.png")

```

</details>

Reply:
I still think it's useful to keep this around. Especially since we want it to broadcast when multiple vectors are passed.

Reply:
Then we keep it 👍